lib/gateway/get-query-result.js

```diff
@@ -0,0 +1,140 @@
+'use strict'
+
+const { preGatewayExecutionHandler } = require('../handlers')
+
+const mergeQueries = (queries) => {
+  const q = [...new Set(queries.map(q => q.query))]
+  const resultIndexes = []
+  const mergedQueries = queries.reduce((acc, curr, queryIndex) => {
+    if (!acc[curr.query]) {
+      acc[curr.query] = {
+        document: curr.document,
+        variables: curr.variables
+      }
+      resultIndexes[q.indexOf(curr.query)] = []
+    } else {
+      acc[curr.query].variables.representations = [
+        ...acc[curr.query].variables.representations,
+        ...curr.variables.representations
+      ]
+    }
+
+    for (let i = 0; i < curr.variables.representations.length; i++) {
+      resultIndexes[q.indexOf(curr.query)].push(queryIndex)
+    }
+
+    return acc
+  }, {})
+
+  return { mergedQueries, resultIndexes }
+}
+
+const getBactchedResult = async ({ mergeQueriesResult, context, serviceDefinition, service }) => {
+  const { mergedQueries, resultIndexes } = mergeQueriesResult
+  const batchedQueries = []
+
+  for (const [query, { document, variables }] of Object.entries(mergedQueries)) {
+    let modifiedQuery
+
+    if (context.preGatewayExecution !== null) {
+      ({ modifiedQuery } = await preGatewayExecutionHandler({
+        schema: serviceDefinition.schema,
+        document,
+        context,
+        service: { name: service }
+      }))
+    }
+
+    batchedQueries.push({
+      operationName: document.definitions.find(d => d.kind === 'OperationDefinition').name.value,
+      query: modifiedQuery || query,
+      variables
+    })
+  }
+
+  const response = await serviceDefinition.sendRequest({
+    originalRequestHeaders: context.reply.request.headers,
+    body: JSON.stringify(batchedQueries),
+    context
+  })
+
+  return buildResult({ resultIndexes, data: response.json })
+}
+
+const buildResult = ({ resultIndexes, data }) => {
+  const result = []
+
+  for (const [queryIndex, queryResponse] of data.entries()) {
+    let entityIndex = 0
+
+    for (const entity of queryResponse.data._entities) {
+      if (!result[resultIndexes[queryIndex][entityIndex]]) {
+        result[resultIndexes[queryIndex][entityIndex]] = {
+          ...queryResponse,
+          json: {
+            data: {
+              _entities: [entity]
+            }
+          }
+        }
+      } else {
+        result[resultIndexes[queryIndex][entityIndex]].json.data._entities.push(entity)
+      }
+
+      entityIndex++
+    }
+  }
+
+  return result
+}
+
+const getResult = async ({ mergeQueriesResult, serviceDefinition, context, service }) => {
```

Comment:
Using named function could help in case of errors being thrown

```suggestion
async function getResult ({ mergeQueriesResult, serviceDefinition, context, service }) {
```